File: 지은/Lv.2/올바른괄호.py
def solution(s):
    res = 1
    if s[0] == ")" or s[-1] == "(":
        return False
    stack = []
    
    for i in s:
        stack.append(i)
        if len(stack)>1:
            if stack[-2] == "(" and stack[-1] == ")":
                stack.pop()
                stack.pop()
    return len(stack) == 0


# 다른 사람의 풀이
def solution2(s):
    pair = 0
    for x in s:
        if pair < 0: break
        if x == "(":
            pair += 1
        else:
            if x== ")":
                pair -= 1
    return pair == 0

File: 성재/Lv.2/올바른_괄호.java
/*
문제 링크 : https://school.programmers.co.kr/learn/courses/30/lessons/12909
 */

import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;

        Stack<Character> stack = new Stack<>();
        char[] char_arr = s.toCharArray();

        // 총 괄호 개수가 짝수가 아니면 false
        if (char_arr.length % 2 != 0) {
            answer = false;
        }

        for (int char_idx : char_arr) {
            if (char_arr[char_idx] == '(') {
                stack.add(char_arr[char_idx]);

            } else if (char_arr[char_idx] == ')') {
                if (!stack.empty()) {
                    stack.pop();
                }
            }
        }

        if (!stack.empty()) {
            answer = false;
        }

        return answer;
    }
}
